container/container.go

```diff
@@ -1,105 +1,243 @@
+	if err != nil {
+		return nil, err
+	}
+	client.NegotiateAPIVersion(context.Background())
+
 	c := &DockerContainer{
-		callTimeout:            60 * time.Second,
-		defaultStopGracePeriod: 10 * time.Second, // default docker value
-		nsprefix:               nsprefix,
+		client:   client,
+		nsprefix: nsprefix,
+	}
+
+	if err := c.isSwarmInit(); err != nil {
+		return nil, err
+	}
+
+	if err := c.createSharedNetwork(); err != nil {
+		return nil, err
 	}
-	for _, option := range options {
-		option(c)
+	return c, nil
+}
+
+// Build builds a docker image.
+func (c *DockerContainer) Build(path string) (tag string, err error) {
+	excludeFiles, err := readDockerIgnoreFile(path)
+	if err != nil {
+		return "", err
+	}
+
+	tr, err := archive.TarWithOptions(path, &archive.TarOptions{
+		Compression:     archive.Gzip,
+		ExcludePatterns: excludeFiles,
+	})
+	if err != nil {
+		return "", err
+	}
+	defer tr.Close()
+
+	resp, err := c.client.ImageBuild(context.Background(), tr, types.ImageBuildOptions{
+		Remove:         true,
+		ForceRemove:    true,
+		SuppressOutput: true,
+	})
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+	return tagFromResponse(resp.Body)
+}
+
+// CreateNetwork creates a Docker Network with a namespace. Retruns network id and error.
+func (c *DockerContainer) CreateNetwork(namespace string) (string, error) {
+	namespace = c.namespace(namespace)
+	network, err := c.client.NetworkInspect(context.Background(), namespace, types.NetworkInspectOptions{})
+	if err != nil && !client.IsErrNotFound(err) {
+		return "", err
 	}
-	var err error
-	if c.client == nil {
-		c.client, err = docker.NewClientWithOpts(docker.FromEnv)
+	if network.ID != "" {
+		return network.ID, nil
+	}
+	response, err := c.client.NetworkCreate(context.Background(), namespace, types.NetworkCreate{
+		CheckDuplicate: true,
+		Driver:         "overlay",
+		Labels: map[string]string{
+			"com.docker.stack.namespace": namespace,
+		},
+	})
+	return response.ID, err
+}
+
+// DeleteNetwork deletes a Docker Network associated with a namespace.
+func (c *DockerContainer) DeleteNetwork(namespace string) error {
+	for {
+		network, err := c.client.NetworkInspect(context.Background(), c.namespace(namespace), types.NetworkInspectOptions{})
+		if client.IsErrNotFound(err) {
+			return nil
+		}
 		if err != nil {
-			return c, err
+			return err
 		}
+		c.client.NetworkRemove(context.Background(), network.ID)
+		time.Sleep(100 * time.Millisecond)
 	}
-	c.negotiateAPIVersion()
-	if err := c.isSwarmInit(); err != nil {
-		return c, err
+}
+
+// SharedNetworkID returns the ID of the shared network.
+func (c *DockerContainer) SharedNetworkID() string {
+	return c.sharedNetworkID
+}
+
+// StartService starts a docker service.
+func (c *DockerContainer) StartService(options ServiceOptions) (string, error) {
+	if status, _ := c.serviceStatus(options.Namespace); status == RUNNING {
```

Comment:
No error management of `serviceStatus` here?

Reply:
No, because only check needed here if it running.